src/ripple/rpc/impl/KeypairForSignature.cpp

```diff
@@ -65,31 +65,43 @@ std::pair<PublicKey, SecretKey>
 keypairForSignature (Json::Value const& params, Json::Value& error)
 {
     bool const has_key_type  = params.isMember (jss::key_type);
-    bool const hasPassphrase = params.isMember (jss::passphrase);
-    bool const hasSecret     = params.isMember (jss::secret);
-    bool const hasSeed       = params.isMember (jss::seed);
-    bool const hasSeedHex    = params.isMember (jss::seed_hex);
 
-    int const n_secrets =
-        (hasPassphrase ? 1 : 0) +
-        (hasSecret ? 1 : 0) +
-        (hasSeed ? 1 : 0) +
-        (hasSeedHex ? 1 : 0);
+    // All of the signature types we allow.  Only one, though!
+    static constexpr char const* const sigTypes[]
+    {
+        jss::passphrase.c_str(),
+        jss::secret.c_str(),
+        jss::seed.c_str(),
+        jss::seed_hex.c_str()
+    };
+
+    // Identify which signature type is in use.
+    constexpr char noSig[] = "no signature";
+    char const* sigType = noSig;
+    int sigTypeCount = 0;
+    for (auto t : sigTypes)
+    {
+        if (params.isMember (t))
+        {
+            ++sigTypeCount;
+            sigType = t;
+        }
+    }
 
-    if (n_secrets == 0)
+    if (sigTypeCount == 0 || sigType == noSig)
```

Comment:
I definitely hear your point about magic strings / numbers.  My take was that the magic was pretty contained here in just this function.  The string "no signature" never escapes unless there's a coding error.  It didn't seem so magical to me.  But it does to you and that matters.

I'm not keen on the `boost::optional`, since dereferencing it can also result in undefined behavior, just like the `nullptr`.  The `vector` approach seems expensive, since we have to go to the heap.